pkg/controller/manager/vlanconfig/controller.go

```diff
@@ -56,62 +57,115 @@ func (h Handler) OnChange(key string, vc *networkv1.VlanConfig) (*networkv1.Vlan
 	if err := h.ensureClusterNetwork(vc.Spec.ClusterNetwork); err != nil {
 		return nil, err
 	}
-	return nil, nil
+	return vc, nil
 }
 
-func (h Handler) OnRemove(key string, vc *networkv1.VlanConfig) (*networkv1.VlanConfig, error) {
-	klog.Infof("vlan config %s has been removed", vc.Name)
-	// delete clusternetwork if there isn't any other vlanconfigs
-	if err := h.deleteClusterNetwork(vc.Spec.ClusterNetwork); err != nil {
-		return nil, err
+func (h Handler) SetClusterNetworkReady(key string, vs *networkv1.VlanStatus) (*networkv1.VlanStatus, error) {
+	if vs == nil || vs.DeletionTimestamp != nil {
+		return nil, nil
 	}
 
-	return vc, nil
+	klog.Infof("vlan status %s has been changed, node: %s, clusterNetwork: %s, vc: %s", vs.Name, vs.Status.Node,
+		vs.Status.ClusterNetwork, vs.Status.VlanConfig)
+
+	if err := h.setClusterNetworkReady(vs); err != nil {
+		return nil, fmt.Errorf("set cluster network of vs %s ready failed, error: %w", vs.Name, err)
+	}
+
+	return vs, nil
+}
+
+func (h Handler) SetClusterNetworkUnready(key string, vs *networkv1.VlanStatus) (*networkv1.VlanStatus, error) {
+	if vs == nil {
+		return nil, nil
+	}
+
+	if err := h.setClusterNetworkUnready(vs); err != nil {
+		return nil, fmt.Errorf("set cluster network unready before deleting vs %s failed, error: %w", vs.Name, err)
+	}
+
+	return vs, nil
 }
 
 func (h Handler) ensureClusterNetwork(name string) error {
-	_, err := h.cnCache.Get(name)
-	if err != nil && !apierrors.IsNotFound(err) {
+	if _, err := h.cnCache.Get(name); err != nil && !apierrors.IsNotFound(err) {
 		return err
 	} else if err == nil {
 		return nil
 	}
+
 	// if cn is not existing
-	if _, err := h.cnClient.Create(&networkv1.ClusterNetwork{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: name,
-		},
-	}); err != nil {
+	cn := &networkv1.ClusterNetwork{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	if _, err := h.cnClient.Create(cn); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (h Handler) deleteClusterNetwork(name string) error {
-	vcs, err := h.vcCache.List(labels.Set(map[string]string{
-		utils.KeyClusterNetworkLabel: name,
-	}).AsSelector())
+func (h Handler) setClusterNetworkReady(vs *networkv1.VlanStatus) error {
+	cn, err := h.cnCache.Get(vs.Status.ClusterNetwork)
 	if err != nil {
 		return err
 	}
 
-	if len(vcs) != 1 {
+	if networkv1.Ready.IsTrue(cn.Status) {
 		return nil
 	}
-	return h.cnClient.Delete(name, &metav1.DeleteOptions{})
+	cnCopy := cn.DeepCopy()
+	networkv1.Ready.True(&cnCopy.Status)
```

Comment:
when network's status is same as the desired, avoid update

Reply:
We check it in line 114.